apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/Transaction.java

```diff
@@ -221,9 +239,63 @@ public void incrementReferences() {
 
     public void decrementReferences() {
         final int referenceCount = this.references.decrementAndGet();
-        super.decrementReferences();
+        if (logger.isTraceEnabled()) {
+            logger.trace("decrement references to {} ({})", this, referenceCount);
+        }
         if (referenceCount == 0) {
             tracer.recycle(this);
         }
     }
+
+    void incrementTimer(@Nullable String type, @Nullable String subtype, long duration) {
+        if (!collectBreakdownMetrics || type == null || finished) {
+            return;
+        }
+        final Labels.Mutable spanType = labelsThreadLocal.get();
+        spanType.resetState();
+        Timer timer = spanTimings.get(spanType.spanType(type).spanSubType(subtype));
+        if (timer == null) {
+            timer = new Timer();
+            Timer racyTimer = spanTimings.putIfAbsent(spanType.immutableCopy(), timer);
+            if (racyTimer != null) {
+                timer = racyTimer;
+            }
+        }
+        timer.update(duration);
+        if (finished) {
+            // in case end()->trackMetrics() has been called concurrently
+            // don't leak timers
+            timer.resetState();
+        }
+    }
+
+    private void trackMetrics() {
+        final String type = getType();
+        if (type == null) {
+            return;
+        }
+        final Labels.Mutable labels = labelsThreadLocal.get();
+        labels.resetState();
+        labels.transactionName(name).transactionType(type);
+        final MetricRegistry metricRegistry = tracer.getMetricRegistry();
+        long criticalValueAtEnter = metricRegistry.writerCriticalSectionEnter();
+        try {
+            metricRegistry.updateTimer("transaction.duration", labels, getDuration());
+            if (collectBreakdownMetrics) {
+                metricRegistry.incrementCounter("transaction.breakdown.count", labels);
+                List<Labels> keyList = spanTimings.keyList();
+                for (int i = 0; i < keyList.size(); i++) {
+                    Labels spanType = keyList.get(i);
```

Comment:
It can't - at least when setting `finished= true` before tracking the metrics.

Reply:
Yes, moving `finished=true` assignment to before `doEnd` should fix that 👍

Reply:
hm, there's still a possibility of a race condition then:

| Thread | action| finished| 
|--------|-|-------|
| B| calls `transaction.incrementTimer`| false |
| A | set finished = true | true |
| B | has already passed the finished check and updates the timer map | true|
| A | at the time the timer is read, only the `totalTime` field has been updated by thread B. `count` is still `0`. | true |

->  `incrementTimer` has to be executed within a critical section and `finished` can only be set to `true` once there is no one in the critical section.

Reply:
There are two issues here- this conversation is about the thread safety of the usage of `spanTimings`, which should be achieved by moving the `finished` flag setting (so that `keyList` won't change while iterating over it).

The other is thread safety of the Timers, as I referred separately in https://github.com/elastic/apm-agent-java/pull/588/files#r294697740, https://github.com/elastic/apm-agent-java/pull/588/files#r294701992 and https://github.com/elastic/apm-agent-java/pull/588/files#r294707481. You can achieve Timer atomicity with critical sections, but I think it is better to make it atomic inherently, so that it can't be written to or read from concurrently (then in race conditions you read full state - either set or not-yet-set, which is OK). In addition, you would still need to make sure it is reset properly either way.